subsys/bluetooth/services/cgms/cgms.c

```diff
@@ -0,0 +1,451 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr/types.h>
+#include <stddef.h>
+#include <string.h>
+#include <errno.h>
+#include <zephyr/zephyr.h>
+#include <zephyr/init.h>
+#include <zephyr/net/buf.h>
+
+#include <zephyr/bluetooth/bluetooth.h>
+#include <zephyr/bluetooth/hci.h>
+#include <zephyr/bluetooth/conn.h>
+#include <zephyr/bluetooth/uuid.h>
+#include <zephyr/bluetooth/gatt.h>
+#include <bluetooth/services/cgms.h>
+#include "cgms_internal.h"
+
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(cgms, CONFIG_BT_CGMS_LOG_LEVEL);
+
+#define CGMS_SVC_MEAS_ATTR_IDX			1
+#define CGMS_SVC_RACP_ATTR_IDX			12
+#define CGMS_SVC_SOCP_ATTR_IDX			15
+
+static struct k_work_delayable report_meas_work;
+
+/* Below are the global variables that describe the state of the CGMS. */
+
+/* The start time of current session. */
+static struct bt_cgm_session_start_time_t	sst;
+/* The session run time, showing how many hours it lasts. */
+static uint16_t					srt;
+/* Sensor's local time. */
+static uint32_t					local_start_time;
+/* Sensor's current status */
+static struct bt_cgm_status_t			status;
+/* Sensor's feature. */
+static struct bt_cgm_feature_t			feature;
+/* The communication interval between sensor and user. */
+uint16_t					comm_interval;
+
+static void cgmc_ccc_cfg_changed(const struct bt_gatt_attr *attr, uint16_t value);
+static ssize_t read_feature(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+				void *buf, uint16_t len, uint16_t offset);
+static ssize_t read_status(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+				void *buf, uint16_t len, uint16_t offset);
+static ssize_t read_session_start_time(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+				void *buf, uint16_t len, uint16_t offset);
+static ssize_t write_session_start_time(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+				const void *buf, uint16_t len, uint16_t offset, uint8_t flags);
+static ssize_t read_session_run_time(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+				void *buf, uint16_t len, uint16_t offset);
+static ssize_t racp_on_receive(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+				const void *buf, uint16_t len, uint16_t offset, uint8_t flags);
+static ssize_t socp_on_receive(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+				const void *buf, uint16_t len, uint16_t offset, uint8_t flags);
+
+BT_GATT_SERVICE_DEFINE(cgms_svc,
```

Comment:
Could you move this definition to the bottom of this file to remove the necessity for declarations above (L48-L62)?

Reply:
As inside these callbacks the struct cgms_svc is used, I have to declare them first, define the gatt service next, and then I can use cgms_svc in the definition of callback function.

Reply:
OK. I get your point. Have changed the order so that seperate declration is eliminated.